diff -ru a/src/block.c b/src/block.c
--- a/src/block.c	2026-09-12 15:22:13.411964500 -0400
+++ b/src/block.c	2026-09-12 15:22:54.289171200 -0400
@@ -45,11 +45,13 @@
 zend_object* php_cmark_node_code_block_create(zend_class_entry *ce) {
 	php_cmark_node_code_block_t *n = 
 		(php_cmark_node_code_block_t*) 
-			ecalloc(1, sizeof(php_cmark_node_code_block_t));
+			ecalloc(1, sizeof(php_cmark_node_code_block_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h.h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h.h), ce);
+
 	n->h.h.std.handlers = &php_cmark_node_code_block_handlers;
 
 	return php_cmark_node_zend(&n->h.h);
diff -ru a/src/custom.c b/src/custom.c
--- a/src/custom.c	2026-09-12 15:22:14.122071500 -0400
+++ b/src/custom.c	2026-09-12 15:22:54.289171200 -0400
@@ -32,11 +32,13 @@
 zend_object* php_cmark_node_custom_create(zend_class_entry *ce) {
 	php_cmark_node_custom_t *n = 
 		(php_cmark_node_custom_t*) 
-			ecalloc(1, sizeof(php_cmark_node_custom_t));
+			ecalloc(1, sizeof(php_cmark_node_custom_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_custom_handlers;
 
 	return php_cmark_node_zend(&n->h);
diff -ru a/src/heading.c b/src/heading.c
--- a/src/heading.c	2026-09-12 15:22:13.558273500 -0400
+++ b/src/heading.c	2026-09-12 15:22:54.306319100 -0400
@@ -41,11 +41,13 @@
 zend_object* php_cmark_node_heading_create(zend_class_entry *ce) {
 	php_cmark_node_heading_t *n = 
 		(php_cmark_node_heading_t*) 
-			ecalloc(1, sizeof(php_cmark_node_heading_t));
+			ecalloc(1, sizeof(php_cmark_node_heading_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_heading_handlers;
 
 	return php_cmark_node_zend(&n->h);
diff -ru a/src/list.c b/src/list.c
--- a/src/list.c	2026-09-12 15:22:13.249523800 -0400
+++ b/src/list.c	2026-09-12 15:22:54.302323800 -0400
@@ -51,11 +51,13 @@
 zend_object* php_cmark_node_list_create(zend_class_entry *ce) {
 	php_cmark_node_list_t *n = 
 		(php_cmark_node_list_t*) 
-			ecalloc(1, sizeof(php_cmark_node_list_t));
+			ecalloc(1, sizeof(php_cmark_node_list_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_list_handlers;
 
 	return php_cmark_node_zend(&n->h);
@@ -228,11 +230,13 @@
 zend_object* php_cmark_node_ordered_list_create(zend_class_entry *ce) {
 	php_cmark_node_ordered_list_t *n = 
 		(php_cmark_node_ordered_list_t*) 
-			ecalloc(1, sizeof(php_cmark_node_ordered_list_t));
+			ecalloc(1, sizeof(php_cmark_node_ordered_list_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_ordered_list_handlers;
 
 	return php_cmark_node_zend(&n->h);
diff -ru a/src/media.c b/src/media.c
--- a/src/media.c	2026-09-12 15:22:13.951976700 -0400
+++ b/src/media.c	2026-09-12 15:22:54.306319100 -0400
@@ -43,11 +43,13 @@
 zend_object* php_cmark_node_media_create(zend_class_entry *ce) {
 	php_cmark_node_media_t *n = 
 		(php_cmark_node_media_t*) 
-			ecalloc(1, sizeof(php_cmark_node_media_t));
+			ecalloc(1, sizeof(php_cmark_node_media_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_media_handlers;
 
 	return php_cmark_node_zend(&n->h);
diff -ru a/src/node.c b/src/node.c
--- a/src/node.c	2026-09-12 15:22:13.017934300 -0400
+++ b/src/node.c	2026-09-12 15:22:54.304320100 -0400
@@ -152,11 +152,13 @@
 zend_object* php_cmark_node_create(zend_class_entry *ce) {
 	php_cmark_node_t *n = 
 		(php_cmark_node_t*) 
-			ecalloc(1, sizeof(php_cmark_node_t));
+			ecalloc(1, sizeof(php_cmark_node_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init
 		(php_cmark_node_zend(n), ce);
 
+	object_properties_init(php_cmark_node_zend(n), ce);
+
 	n->std.handlers = &php_cmark_node_handlers;
 
 	return php_cmark_node_zend(n);
diff -ru a/src/text.c b/src/text.c
--- a/src/text.c	2026-09-12 15:22:13.716188000 -0400
+++ b/src/text.c	2026-09-12 15:22:54.305319500 -0400
@@ -43,11 +43,13 @@
 zend_object* php_cmark_node_text_create(zend_class_entry *ce) {
 	php_cmark_node_text_t *n = 
 		(php_cmark_node_text_t*) 
-			ecalloc(1, sizeof(php_cmark_node_text_t));
+			ecalloc(1, sizeof(php_cmark_node_text_t) + zend_object_properties_size(ce));
 
 	zend_object_std_init(
 		php_cmark_node_zend(&n->h), ce);
 
+	object_properties_init(php_cmark_node_zend(&n->h), ce);
+
 	n->h.std.handlers = &php_cmark_node_text_handlers;
 
 	return php_cmark_node_zend(&n->h);
