diff -ru a/src/block.c b/src/block.c
--- a/src/block.c	2026-09-12 13:57:22.523315200 -0400
+++ b/src/block.c	2026-09-12 13:57:23.142019200 -0400
@@ -278,7 +278,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "CodeBlock", php_cmark_node_code_block_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_code_block_ce = zend_register_internal_class_ex(&ce, php_cmark_node_text_ce);
+	php_cmark_node_code_block_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_code_block_ce->create_object = php_cmark_node_code_block_create;
 
 #define php_cmark_node_code_block_property(n) \
@@ -296,11 +298,15 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "HTMLBlock", php_cmark_node_html_block_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_html_block_ce = zend_register_internal_class_ex(&ce, php_cmark_node_text_ce);
+	php_cmark_node_html_block_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "CustomBlock", php_cmark_node_custom_block_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_custom_block_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_custom_block_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_custom_block_ce->create_object = php_cmark_node_custom_create;
 
 #define php_cmark_node_custom_block_property(n) \
diff -ru a/src/break.c b/src/break.c
--- a/src/break.c	2026-09-12 13:57:22.741402900 -0400
+++ b/src/break.c	2026-09-12 13:57:23.156050300 -0400
@@ -70,15 +70,21 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "ThematicBreak", php_cmark_node_thematic_break_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_thematic_break_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_thematic_break_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "SoftBreak", php_cmark_node_soft_break_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_soft_break_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_soft_break_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "LineBreak", php_cmark_node_line_break_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_line_break_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_line_break_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/code.c b/src/code.c
--- a/src/code.c	2026-09-12 13:57:22.887493000 -0400
+++ b/src/code.c	2026-09-12 13:57:23.168180400 -0400
@@ -64,7 +64,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Code", php_cmark_node_code_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_code_ce = zend_register_internal_class_ex(&ce, php_cmark_node_text_ce);
+	php_cmark_node_code_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/document.c b/src/document.c
--- a/src/document.c	2026-09-12 13:57:22.205111700 -0400
+++ b/src/document.c	2026-09-12 13:57:23.142019200 -0400
@@ -43,7 +43,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Document", php_cmark_node_document_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_document_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_document_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/heading.c b/src/heading.c
--- a/src/heading.c	2026-09-12 13:57:22.673693300 -0400
+++ b/src/heading.c	2026-09-12 13:57:23.156050300 -0400
@@ -194,7 +194,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Heading", php_cmark_node_heading_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_heading_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_heading_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_heading_ce->create_object = php_cmark_node_heading_create;
 
 #define php_cmark_node_heading_property(n) \
diff -ru a/src/inline.c b/src/inline.c
--- a/src/inline.c	2026-09-12 13:57:22.966980100 -0400
+++ b/src/inline.c	2026-09-12 13:57:23.168180400 -0400
@@ -78,11 +78,15 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "HTMLInline", php_cmark_node_html_inline_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_html_inline_ce = zend_register_internal_class_ex(&ce, php_cmark_node_text_ce);
+	php_cmark_node_html_inline_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "CustomInline", php_cmark_node_custom_inline_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_custom_inline_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_custom_inline_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_custom_inline_ce->create_object = php_cmark_node_custom_create;
 
 #define php_cmark_node_custom_inline_property(n) \
diff -ru a/src/item.c b/src/item.c
--- a/src/item.c	2026-09-12 13:57:22.437711100 -0400
+++ b/src/item.c	2026-09-12 13:57:23.142019200 -0400
@@ -43,7 +43,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Item", php_cmark_node_item_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_item_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_item_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/list.c b/src/list.c
--- a/src/list.c	2026-09-12 13:57:22.365971900 -0400
+++ b/src/list.c	2026-09-12 13:57:23.142019200 -0400
@@ -443,7 +443,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "OrderedList", php_cmark_node_list_ordered_methods);
 	
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_list_ordered_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_list_ordered_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_list_ordered_ce->create_object = php_cmark_node_ordered_list_create;
 
 #define php_cmark_node_list_property(n) \
@@ -462,7 +464,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "BulletList", php_cmark_node_list_bullet_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_list_bullet_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_list_bullet_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_list_bullet_ce->create_object = php_cmark_node_list_create;
 
 #define php_cmark_node_list_property(n) \
diff -ru a/src/media.c b/src/media.c
--- a/src/media.c	2026-09-12 13:57:23.049514100 -0400
+++ b/src/media.c	2026-09-12 13:57:23.168180400 -0400
@@ -317,7 +317,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Link", php_cmark_node_link_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_link_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_link_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_link_ce->create_object = php_cmark_node_media_create;
 
 #define php_cmark_node_link_property(n) \
@@ -329,7 +331,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Image", php_cmark_node_image_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_image_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_image_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_image_ce->create_object = php_cmark_node_media_create;
 
 #define php_cmark_node_image_property(n) \
diff -ru a/src/paragraph.c b/src/paragraph.c
--- a/src/paragraph.c	2026-09-12 13:57:22.598893800 -0400
+++ b/src/paragraph.c	2026-09-12 13:57:23.156050300 -0400
@@ -43,7 +43,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Paragraph", php_cmark_node_paragraph_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_paragraph_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_paragraph_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/quote.c b/src/quote.c
--- a/src/quote.c	2026-09-12 13:57:22.294924200 -0400
+++ b/src/quote.c	2026-09-12 13:57:23.142019200 -0400
@@ -43,7 +43,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "BlockQuote", php_cmark_node_quote_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_quote_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_quote_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
diff -ru a/src/text.c b/src/text.c
--- a/src/text.c	2026-09-12 13:57:22.820776200 -0400
+++ b/src/text.c	2026-09-12 13:57:23.156050300 -0400
@@ -240,7 +240,9 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node", "Text", php_cmark_node_text_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_text_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_text_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_text_ce->create_object = php_cmark_node_text_create;
 
 #define php_cmark_node_text_property(n) \
@@ -259,11 +261,15 @@
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node\\Text", "Emphasis", php_cmark_node_text_emphasis_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_text_emphasis_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_text_emphasis_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	INIT_NS_CLASS_ENTRY(ce, "CommonMark\\Node\\Text", "Strong", php_cmark_node_text_strong_methods);
 
+	ce.ce_flags |= ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 	php_cmark_node_text_strong_ce = zend_register_internal_class_ex(&ce, php_cmark_node_ce);
+	php_cmark_node_text_strong_ce->ce_flags &= ~ZEND_ACC_EXPLICIT_ABSTRACT_CLASS;
 
 	return SUCCESS;
 }
